src/main/java/io/nats/client/impl/NatsConnection.java

```diff
@@ -55,44 +56,45 @@
                                                     // connecting
 
     private Status status;
-    private ReentrantLock statusLock;
-    private Condition statusChanged;
+    private final ReentrantLock statusLock;
+    private final Condition statusChanged;
 
     private CompletableFuture<DataPort> dataPortFuture;
     private DataPort dataPort;
     private String currentServerURI;
     private CompletableFuture<Boolean> reconnectWaiter;
-    private HashMap<String, String> serverAuthErrors;
+    private final HashMap<String, String> serverAuthErrors;
 
-    private NatsConnectionReader reader;
-    private NatsConnectionWriter writer;
+    private final NatsConnectionReader reader;
+    private final NatsConnectionWriter writer;
 
-    private AtomicReference<ServerInfo> serverInfo;
+    private final AtomicReference<ServerInfo> serverInfo;
 
-    private Map<String, NatsSubscription> subscribers;
-    private Map<String, NatsDispatcher> dispatchers; // use a concurrent map so we get more consistent iteration
+    private final Map<String, NatsSubscription> subscribers;
+    private final Map<String, NatsDispatcher> dispatchers; // use a concurrent map so we get more consistent iteration
                                                      // behavior
-    private Map<String, CompletableFuture<Message>> responses;
-    private ConcurrentLinkedDeque<CompletableFuture<Boolean>> pongQueue;
+    private final Map<String, NatsRequestCompletableFuture> responsesAwaiting;
+    private final Map<String, NatsRequestCompletableFuture> responsesRespondedTo;

```

Comment:
responses changed to responsesAwaiting and responsesRespondedTo. Awaiting is the same as the original. Responded To gives a places to track after the first response